torch/csrc/jit/passes/batch_mm.cpp

```diff
@@ -201,6 +201,48 @@ struct TreeToken {
     }
     return matmuls;
   }
+
+  // Checks whether a tree merge is safe by trying to relocate all nodes in the
+  // tree together.
+  bool canMergeTree(const AliasDb& db) {
+    std::vector<Node*> queue{node};
+    while (!queue.empty()) {
+      auto n = queue.back();
+      queue.pop_back();
+      if (n->matches("aten::t(Tensor self) -> Tensor")) {
+        Node* input_node = n->input()->node();
+        JIT_ASSERT(input_node->matches(
+            "aten::mm(Tensor self, Tensor mat2) -> Tensor"));
+        if (!n->moveAfterTopologicallyValid(input_node, db)) {
+          return false;
+        }
+      } else if (
+          n->matches(
+              "aten::add(Tensor self, Tensor other, *, Scalar alpha) -> Tensor")) {
+        auto a = n->inputs()[0]->node();
+        auto b = n->inputs()[1]->node();
+        bool couldMove = true;
+        if (a->isAfter(b)) {
```

Comment:
It shouldn't depend what their order is, because they are guaranteed that they are independent now that we've got here.